### PR TITLE
fixed user's own message is `unread` on different plaforms 

### LINKED ISSUE #1985: User's own messages displaying as unread in other platforms
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

## Expected Result:
When a user sends a message in one platform (e.g., Desktop) using Expensify.cash, this user's own message should not show up as unread in another platform (e.g., Mobile).

## Actual Result:
The user's own messages sent in one platform show up as unread in the other platform (e.g., messages sent in desktop show as unread in mobile web and vice versa).

## Action Performed:

1. Sign into the Expensify.Cash desktop app
1. Sign into Expensify.Cash on the mobile app (or any other platform, eg: web, mobile web)
1. Send a message to a user from the desktop app
1. Switch to the mobile app (or whatever second platform you're using) and observe that your own message is marked as unread
1. Send a message to a user from the mobile app (or whatever second platform you're using)
1. Switch to the desktop app and observe that your own message is marked as unread

## Workaround:
The workaround is to click into the threads to mark them as read.

## Platform:
Web: Yes
iOS: Yes
Android: Unknown
Desktop App: Yes
Mobile Web: Yes

**Version Number:** v1.0.2-73
**Notes/Photos/Videos:** Any additional supporting documentation

image

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/157742

View the job on Upwork [here](https://www.upwork.com/jobs/~01fde57562f1db3f1b).

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review @marcaaron 

### Details
 we were using the stale reference to `lastReadSequenceNumber` which caused the linked issue. I have refactored the code and fixed it.

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #1985 

### Tests
Tried sending messages as explained in the issue.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [ ] iOS
- [x] Android

https://user-images.githubusercontent.com/24370807/112392837-183c3180-8d20-11eb-8099-781008fe3169.mp4

---
> ### To another User

  https://user-images.githubusercontent.com/24370807/112392850-1ecaa900-8d20-11eb-9ca3-8050ee8d3013.mp4

### CODE REVIEW

File: src/libs/actions/Report.js [hunk 1]
```diff
@@ -364,6 +355,19 @@ function updateReportWithNewAction(reportID, reportAction) {
...
 
+    // When handling an action from the current users we can assume that their
+    // last read actionID has been updated in the server but not necessarily reflected
+    // locally so we must first update it and then calculate the unread (which should be 0)
+    if (isFromCurrentUser) {
+        setLocalLastRead(reportID, newMaxSequenceNumber);
+    }
+
+    // We update the lastReadSequenceNumber to reflect new.
+    lastReadSequenceNumber = lastReadSequenceNumbers[reportID] || 0;
```
Comment (marcaaron) [inline src/libs/actions/Report.js:366]: There's no need to reassign this unless it's from the current user so we can push it into the `isFromCurrentUser` block

File: src/libs/actions/Report.js [hunk 2]
```diff
@@ -364,6 +355,19 @@ function updateReportWithNewAction(reportID, reportAction) {
         updatedReportObject.lastActorEmail = reportAction.actorEmail;
     }
 
+    // When handling an action from the current users we can assume that their
+    // last read actionID has been updated in the server but not necessarily reflected
+    // locally so we must first update it and then calculate the unread (which should be 0)
+    if (isFromCurrentUser) {
+        setLocalLastRead(reportID, newMaxSequenceNumber);
+    }
```
Comment (marcaaron) [inline src/libs/actions/Report.js:363]: This should not need to be moved from where it was before... unless I am missing something...

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: src/libs/actions/Report.js (hunk 2 above)
Comment (parasharrajat) [reply src/libs/actions/Report.js:363]: Yeah Correct.

File: src/libs/actions/Report.js (hunk 1 above)
Comment (parasharrajat) [reply src/libs/actions/Report.js]: Done

File: src/libs/actions/Report.js (hunk 2 above)
Comment (marcaaron) [reply src/libs/actions/Report.js]: Hmm ok this is kind of confusing and I think likely to break again when someone tries to clean up this logic.

File: src/libs/actions/Report.js [hunk 3]
```diff
@@ -336,23 +336,14 @@ function removeOptimisticActions(reportID) {
...
-    const lastReadSequenceNumber = lastReadSequenceNumbers[reportID] || 0;
-
-    // When handling an action from the current users we can assume that their
-    // last read actionID has been updated in the server but not necessarily reflected
-    // locally so we must first update it and then calculate the unread (which should be 0)
-    if (isFromCurrentUser) {
-        setLocalLastRead(reportID, newMaxSequenceNumber);
-    }
-
+    let lastReadSequenceNumber = lastReadSequenceNumbers[reportID] || 0;
```
Comment (marcaaron) [inline src/libs/actions/Report.js:339]: I think this would be clearer if we set this up as it's own variable. So, something like 

```js
const initialLastReadSequenceNumber = lastReadSequenceNumbers[reportID] || 0;
```

Then rather than re-assign it we should do this check where it was before
```js
    if (isFromCurrentUser) {
        setLocalLastRead(reportID, newMaxSequenceNumber);
    }
```

File: src/libs/actions/Report.js [hunk 4]
```diff
@@ -336,23 +336,14 @@ function removeOptimisticActions(reportID) {
...
     const messageText = lodashGet(reportAction, ['message', 0, 'text'], '');
 
     // Always merge the reportID into Onyx
     // If the report doesn't exist in Onyx yet, then all the rest of the data will be filled out
     // by handleReportChanged
     const updatedReportObject = {
         reportID,
-        unreadActionCount: newMaxSequenceNumber - lastReadSequenceNumber,
```
Comment (marcaaron) [inline src/libs/actions/Report.js:355]: Add this back in here but reference 

```js
lastReadSequenceNumbers[reportID] || 0;
```

directly with a comment that the value may have been modified by `setLocalLastRead()`

**Review by marcaaron — COMMENTED** [review]
Lmk if these comments make sense. I'm a little concerned the existing logic is hard to follow and will break again.

File: src/libs/actions/Report.js (hunk 4 above)
Comment (parasharrajat) [reply src/libs/actions/Report.js:355]: OK. I see. This would be easier to read.

**Review by marcaaron — COMMENTED** [review]
👍

**Review by timszot — APPROVED** [review]

**Review by marcaaron — APPROVED** [review]
LGTM
